### PR TITLE
Add the project artifact as Spark Livy session dependency with deployment

### PR DESCRIPTION
**Pull request by wezhang**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt [hunk 1]
```diff
@@ -83,20 +90,59 @@ open class SparkScalaLivyConsoleRunConfiguration(project: Project,
...
-                cluster.httpUserName,
-                cluster.httpPassword)
+                sparkCluster.httpUserName,
+                sparkCluster.httpPassword,
+                logObserver)
+    }
 
-        return SparkScalaLivyConsoleRunProfileState(consoleBuilder, session)
+    open fun applyRunConfiguration(sparkCluster: IClusterDetail, session: SparkSession) {
+        val batchSubmitModel = this.batchRunConfiguration?.submitModel ?: throw ExecutionException(
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt]: Maybe we can do some parameter validation here like `LivySparkBatchJobRunConfiguration.checkSubmissionConfigurationBeforeRun`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt [hunk 2]
```diff
@@ -49,25 +54,25 @@ class SparkScalaLivyConsoleRunProfileState(private val consoleBuilder: SparkScal
...
         val console = consoleBuilder.console
-
-        val livySessionProcessHandler = SparkLivySessionProcessHandler(SparkLivySessionProcess(session))
+        val progressBarScheduler = IdeaSchedulers(consoleBuilder.project)
+        val livySessionProcess = SparkLivySessionProcess(progressBarScheduler, session)
+        val livySessionProcessHandler = SparkLivySessionProcessHandler(livySessionProcess)
 
         console.attachToProcess(livySessionProcessHandler)
+        logSubject.subscribe({ typedMessage -> console.print(typedMessage) }, { err ->
+            console.print("Livy interactive session is stop due to the error ${err.message}", LOG_ERROR_OUTPUT)
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt]: wording: `Livy interactive session is stopped due to error`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt [hunk 3]
```diff
@@ -49,25 +54,25 @@ class SparkScalaLivyConsoleRunProfileState(private val consoleBuilder: SparkScal
...
+                .start()
                 .subscribe({
                     livySessionProcessHandler.execute(postStartCodes)
                     (console as? ScalaLanguageConsole)?.apply {
                         // Customize the Spark Livy interactive console
                         prompt = "\nSpark>"
                     }
                 }, { ex ->
-                    console.print("Can't start Livy interactive session: ${ex.message}\n", ConsoleViewContentType.LOG_ERROR_OUTPUT)
+                    console.print("Can't start Livy interactive session: ${ex.message}\n", LOG_ERROR_OUTPUT)
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt]: Maybe we can replace `console.print` with `logSubject.onNext`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt (hunk 1 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt]: Good suggestion! I'll try to change the class to inherit `RunProfileStatePrepare`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt (hunk 2 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt]: Fix it!

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt (hunk 3 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunProfileState.kt]: Good suggestion!

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/livy/interactive/Session.java [hunk 4]
```diff
@@ -89,55 +99,66 @@
 
     private String executorMemory = "4G";   // Default executor memory to create
 
+    private List<String> files = new ArrayList<>(); // Files for reference
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/livy/interactive/Session.java]: There are seperated paramter lists defined for batches and sessions in http://livy.incubator.apache.org./docs/latest/rest-api.html. I prefer not merging them in code.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/livy/interactive/Session.java [hunk 5]
```diff
@@ -89,55 +99,66 @@
...
+                   final URI baseUrl,
+                   final @Nullable String username,
+                   final @Nullable String password,
+                   final Observer<SimpleImmutableEntry<MessageInfoType, String>> logObserver) {
         this.name = name;
         this.baseUrl = baseUrl;
         this.lastState = SessionState.NOT_STARTED;
-        this.http = new HttpObservable(username, password);
+
+        if (username == null || password == null) {
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/livy/interactive/Session.java]: Np

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/livy/interactive/Session.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (t-rufang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/livy/interactive/Session.java]: Thanks for the clarification!

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt [hunk 6]
```diff
@@ -83,24 +96,78 @@ open class SparkScalaLivyConsoleRunConfiguration(project: Project,
...
-    override fun checkRunnerSettings(runner: ProgramRunner<*>, runnerSettings: RunnerSettings?, configurationPerRunnerSettings: ConfigurationPerRunnerSettings?) {
-        cluster = ClusterManagerEx.getInstance().getClusterDetailByName(clusterName)
-                .orElseThrow { RuntimeConfigurationError("Can't find the target cluster $clusterName") }
+    override fun checkRunnerSettings(runner: ProgramRunner<*>,
+                                     runnerSettings: RunnerSettings?,
+                                     configurationPerRunnerSettings: ConfigurationPerRunnerSettings?) {
+        batchRunConfiguration?.checkRunnerSettings(runner, runnerSettings, configurationPerRunnerSettings)
+    }
+
+    override fun prepare(runner: ProgramRunner<RunnerSettings>): Observable<SparkScalaLivyConsoleRunConfiguration> {
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt]: Just in curious, where is this `prepare` method get called?

Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/console/SparkScalaLivyConsoleRunConfiguration.kt]: Aligning with Spark batch job, invoked by `RunConfigurationActionUtils::checkAndPrepareRunProfileState()` in `,runEnvironmentProfileWithCheckSettings`

**Pull request merged**